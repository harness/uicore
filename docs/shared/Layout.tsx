import React from 'react'
import { HarnessIcons, Layout, Config } from '../static'
import Nav from './Nav'
import { MDXProvider } from '@mdx-js/react'
import MdxComponents from './MdxComponents'
import { FocusStyleManager } from '@blueprintjs/core'

FocusStyleManager.onlyShowFocusOnTabs()
Config.DISABLE_LINK_REWRITE = true

export default class extends React.Component {
  render() {
    const HarnessLogo: React.ElementType = HarnessIcons['harness-logo-black']

    return (
      <div className="page-container">
        <header>
          <h1>
            <a href="/">
              <HarnessLogo height={20} />
              <span>&nbsp;// UICore</span>
            </a>
          </h1>
        </header>

        <nav>
          <Nav />
        </nav>

        <MDXProvider components={MdxComponents}>
          <main>
            <Layout.Vertical id="main-layout" spacing="medium">
              {this.props.children}
            </Layout.Vertical>
          </main>
        </MDXProvider>

        <style jsx>{`
          .page-container {
            --header-height: 64px;
            --nav-width: 300px;
          }

          header {
            position: sticky;
            height: var(--header-height);
            width: 100%;
            top: 0;
            background-color: var(--white);
            border-bottom: 1px solid var(--grey-200);
            display: flex;
            padding-left: 10px;
            z-index: 1;
          }

          header h1 {
            display: flex;
            align-items: center;
            margin: 0;
            padding: 0;
          }

          header h1 a {
            display: flex;
            align-items: center;
            white-space: nowrap;
          }

          header h1 a:hover {
            text-decoration: none;
          }

          header h1 a span {
            color: var(--grey-400);
            font-weight: 400;
            font-size: 22px;
          }

          nav {
            position: fixed;
            width: var(--nav-width);
            top: 64px;
            left: 0;
            bottom: 0;
            overflow: auto;
          }

          main {
            margin-left: var(--nav-width);
            padding: 20px 20px 100px 20px;
          }

          main :global(pre) {
            margin-top: 0;
          }

          main > :global(h1),
          main > :global(h2),
          main > :global(h3),
          main > :global(h4),
          main > :global(h5),
          main > :global(h6) {
            margin: 30px 0 15px 0;
          }
        `}</style>
      </div>
    )
  }
}
