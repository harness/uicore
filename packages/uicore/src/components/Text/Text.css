/*
 * Copyright (c) Harness Inc.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

.lineclamp {
  display: -webkit-box !important;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: var(--text-line-clamp);
  overflow: hidden;
  word-break: break-all;
}

.withDocsTooltip {
  display: flex;
  align-items: center;
}

.targetClass {
  display: flex !important;
}
.icon {
  vertical-align: middle;
}
