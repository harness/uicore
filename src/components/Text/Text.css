.lineclamp {
  display: -webkit-box !important;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: var(--text-line-clamp);
  overflow: hidden;
  word-break: break-all;
}

.withDocsTooltip {
  display: flex;
  align-items: center;
}

.targetClass {
  display: flex !important;
}
.icon {
  vertical-align: middle;
}
