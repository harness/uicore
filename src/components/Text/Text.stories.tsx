import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary, Description } from '@storybook/addon-docs/blocks'
import { Text, TextProps, Layout, Container, Color, SupText } from '../..'
import { FontVariation } from '../../styled-props/font/FontProps'

export default {
  title: 'Components / Text',

  component: Text,
  subcomponents: { SupText },
  parameters: {
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>Text</Title>
            <Subtitle>
              <pre>
                <code>{`import {Text} from '@harness/uicore'`}</code>
              </pre>
            </Subtitle>
            <Description>{`When lineClamp prop is provided, Text will handle string truncation automatically. Note that the truncation only happens when there is not enough space to render the text content.

`}</Description>
            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [
    Story => (
      <Layout.Vertical>
        <Story />
      </Layout.Vertical>
    )
  ]
} as Meta
export const Basic: Story<TextProps> = () => {
  return (
    <>
      <Text
        font={{ variation: FontVariation.DISPLAY1 }}
        tooltip={`<Text font={{ variation: FontVariation.DISPLAY1 }}>Display 1</Text>`}>
        Display 1
      </Text>
      <Text
        font={{ variation: FontVariation.DISPLAY2 }}
        tooltip={`<Text font={{ variation: FontVariation.DISPLAY2 }}>Display 2</Text>`}>
        Display 2
      </Text>
      <Text
        font={{ variation: FontVariation.H1 }}
        tooltip={`<Text font={{ variation: FontVariation.H1 }}>Headline Text - H1/Bold - Module Landing Title</Text>`}>
        Headline Text - H1/Bold - Module Landing Title
      </Text>
      <Text
        font={{ variation: FontVariation.H1_SEMI }}
        tooltip={`<Text font={{ variation: FontVariation.H1_SEMI }}>Headline Text - H1/Semi - Semi Bold</Text>`}>
        Headline Text - H1/Semi - Semi Bold
      </Text>
      <Text
        font={{ variation: FontVariation.H2 }}
        tooltip={`<Text font={{ variation: FontVariation.H2 }}>Headline Text - H2</Text>`}>
        Headline Text - H2
      </Text>
      <Text
        font={{ variation: FontVariation.H3 }}
        tooltip={`<Text font={{ variation: FontVariation.H3 }}>Headline Text - H3 - Dialogue Title</Text>`}>
        Headline Text - H3 - Dialogue Title
      </Text>
      <Text
        font={{ variation: FontVariation.H4 }}
        tooltip={`<Text font={{ variation: FontVariation.H4 }}>Headline Text - H4 - Page Header Title/Landing/Listing Page Empty State Title</Text>`}>
        Headline Text - H4 - Page Header Title/Landing/Listing Page Empty State Title
      </Text>
      <Text
        font={{ variation: FontVariation.H5 }}
        tooltip={`<Text font={{ variation: FontVariation.H5 }}>Headline Text - H5 - Form/Section Title</Text>`}>
        Headline Text - H5 - Form/Section Title
      </Text>
      <Text
        font={{ variation: FontVariation.H6 }}
        tooltip={`<Text font={{ variation: FontVariation.H6 }}>Headline Text - H6 - Section Title</Text>`}>
        Headline Text - H6 - Section Title
      </Text>
      <Text
        font={{ variation: FontVariation.LEAD }}
        tooltip={`<Text font={{ variation: FontVariation.LEAD }}>Lead</Text>`}>
        Lead
      </Text>
      <Text
        font={{ variation: FontVariation.BODY1 }}
        tooltip={`<Text font={{ variation: FontVariation.BODY1 }}>Body 1 - Landing page body text/Landing page (Empty state body text)</Text>`}>
        Body 1 - Landing page body text/Landing page (Empty state body text){' '}
      </Text>
      <Text
        font={{ variation: FontVariation.BODY2 }}
        tooltip={`<Text font={{ variation: FontVariation.BODY2 }}>Body 2 - Emphasized Text </Text>`}>
        Body 2 - Emphasized Text
      </Text>
      <Text
        font={{ variation: FontVariation.BODY }}
        tooltip={`<Text font={{ variation: FontVariation.BODY }}>Body - Normal Text</Text>`}>
        Body - Normal Text
      </Text>
      <Text
        font={{ variation: FontVariation.BLOCKQUOTE }}
        tooltip={`<Text font={{ variation: FontVariation.BLOCKQUOTE }}>Blockquote</Text>`}>
        Blockquote
      </Text>
      <Text
        font={{ variation: FontVariation.UPPERCASED }}
        tooltip={`<Text font={{ variation: FontVariation.UPPERCASED }}>UPPERCASED (CAPITALIZED on Design)</Text>`}>
        UPPERCASED (CAPITALIZED on Design)
      </Text>
      <Text
        font={{ variation: FontVariation.SMALL_BOLD }}
        tooltip={`<Text font={{ variation: FontVariation.SMALL_BOLD }}>Small Bold</Text>`}>
        Small Bold
      </Text>
      <Text
        font={{ variation: FontVariation.SMALL_SEMI }}
        tooltip={`<Text font={{ variation: FontVariation.SMALL_SEMI }}>Small Semi Bold</Text>`}>
        Small Semi Bold
      </Text>
      <Text
        font={{ variation: FontVariation.SMALL }}
        tooltip={`<Text font={{ variation: FontVariation.SMALL }}>Small - Supplementary info/Breadcrumb</Text>`}>
        Small - Supplementary info/Breadcrumb
      </Text>
      <Text
        font={{ variation: FontVariation.TABLE_HEADERS }}
        tooltip={`<Text font={{ variation: FontVariation.TABLE_HEADERS }}>TABLE HEADERS - LIST/TABLE HEADER (GRAY BG)/LIST/TABLE HEADER (WHITE BG)</Text>`}>
        TABLE HEADERS - LIST/TABLE HEADER (GRAY BG)/LIST/TABLE HEADER (WHITE BG)
      </Text>
      <Text
        font={{ variation: FontVariation.TINY }}
        tooltip={`<Text font={{ variation: FontVariation.TINY }}>Tiny</Text>`}>
        Tiny
      </Text>
      <Text
        font={{ variation: FontVariation.TINY_SEMI }}
        tooltip={`<Text font={{ variation: FontVariation.TINY_SEMI }}>Tiny Semi</Text>`}>
        Tiny Semi
      </Text>
      <Text
        font={{ variation: FontVariation.YAML }}
        tooltip={`<Text font={{ variation: FontVariation.YAML }} tag="pre">YAML - YAML Code</Text>`}
        tag="pre">
        YAML - YAML Code
      </Text>
      <Text
        font={{ variation: FontVariation.CARD_TITLE }}
        tooltip={`<Text font={{ variation: FontVariation.CARD_TITLE }}>Card Title - Project/Execution/Templates/Dashboards</Text>`}>
        Card Title - Project/Execution/Templates/Dashboards
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_TITLE }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_TITLE }}>Form Title</Text>`}>
        Form Title
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_SUB_SECTION }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_SUB_SECTION }}>Form Sub Section</Text>`}>
        Form Sub Section
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_INPUT_TEXT }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_INPUT_TEXT }}>Form Input Text</Text>`}>
        Form Input Text
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_LABEL }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_LABEL }}>Form Label</Text>`}>
        Form Label
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_HELP }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_HELP }}>Form Help Text</Text>`}>
        Form Help Text
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_MESSAGE_DANGER }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_MESSAGE_DANGER }}>Form Message Danger</Text>`}>
        Form Message Danger
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_MESSAGE_WARNING }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_MESSAGE_WARNING }}>Form Message Warning</Text>`}>
        Form Message Warning
      </Text>
      <Text
        font={{ variation: FontVariation.FORM_MESSAGE_SUCCESS }}
        tooltip={`<Text font={{ variation: FontVariation.FORM_MESSAGE_SUCCESS }}>Form Message Success</Text>`}>
        Form Message Success
      </Text>
    </>
  )
}
export const FontWeight: Story<TextProps> = args => {
  return (
    <>
      <Text font={{ weight: 'bold' }} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text font={{ weight: 'semi-bold' }} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text font={{ weight: 'light' }} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </>
  )
}
export const Mono: Story<TextProps> = args => {
  return (
    <>
      <Text font={{ mono: true }} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </>
  )
}
export const Intent: Story<TextProps> = args => {
  return (
    <>
      <Text intent="primary" {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text intent="success" {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text intent="warning" {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text intent="danger" {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </>
  )
}
export const EllipsisTextTruncation: Story<TextProps> = args => {
  return (
    <Layout.Horizontal spacing="medium">
      <Text font="normal" color="red500" lineClamp={1} {...args}>
        The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
        jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy
        dog. The quick brown fox jumps over the lazy dog
      </Text>
      <Text font="normal" color="green500" lineClamp={1} {...args}>
        The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
        jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy
        dog. The quick brown fox jumps over the lazy dog
      </Text>
      <Text font="normal" color="blue500" lineClamp={1} {...args}>
        The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
        jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy
        dog. The quick brown fox jumps over the lazy dog
      </Text>
    </Layout.Horizontal>
  )
}
export const SingleLineWithFixedWidthSetFromParent: Story<TextProps> = args => {
  return (
    <Container width={300} flex>
      <Text font="normal" color="red500" lineClamp={1} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text font="normal" color="green500" lineClamp={1} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
      <Text font="normal" color="blue500" lineClamp={1} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </Container>
  )
}
export const SingleLineWithWidthSetFromTextItself: Story<TextProps> = args => {
  return (
    <>
      <Text font="normal" lineClamp={1} width={100} {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </>
  )
}
export const CustomTooltip: Story<TextProps> = args => {
  return (
    <>
      <Text
        font="normal"
        lineClamp={1}
        width={100}
        tooltip={<Container padding="small">Custom tooltip</Container>}
        {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </>
  )
}
export const AlwaysShowTooltip: Story<TextProps> = args => {
  return (
    <>
      <Text>Note:Always show tooltip (even in case lineClamp is not needed) - tooltip needs to be passed.</Text>
      <Text
        font="normal"
        lineClamp={1}
        width={400}
        tooltip={<Container padding="small">Custom tooltip</Container>}
        alwaysShowTooltip
        {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </>
  )
}
export const MultipleLines: Story<TextProps> = args => {
  return (
    <Text font="normal" lineClamp={2} {...args}>
      The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
      jumps over the lazy dog.The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.
      The quick brown fox jumps over the lazy dog.The quick brown fox jumps over the lazy dog. The quick brown fox jumps
      over the lazy dog. The quick brown fox jumps over the lazy dog.The quick brown fox jumps over the lazy dog. The
      quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.The quick brown fox jumps
      over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.The
      quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps
      over the lazy dog.
    </Text>
  )
}
export const MultipleLineWithFluidWidthFromParent: Story<TextProps> = args => {
  return (
    <>
      <Layout.Horizontal spacing="medium">
        <Text font="normal" color="red500" lineClamp={2} {...args}>
          The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
          jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy
          dog. The quick brown fox jumps over the lazy dog
        </Text>
        <Text font="normal" color="green500" lineClamp={2} {...args}>
          The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
          jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy
          dog. The quick brown fox jumps over the lazy dog
        </Text>
        <Text font="normal" color="blue500" lineClamp={2} {...args}>
          The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
          jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy
          dog. The quick brown fox jumps over the lazy dog
        </Text>
      </Layout.Horizontal>
    </>
  )
}
export const CustomTooltip2: Story<TextProps> = args => {
  return (
    <Text font="normal" lineClamp={2} tooltip={<Container padding="small">Custom tooltip</Container>} {...args}>
      The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
      jumps over the lazy dog.The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.
      The quick brown fox jumps over the lazy dog.The quick brown fox jumps over the lazy dog. The quick brown fox jumps
      over the lazy dog. The quick brown fox jumps over the lazy dog.The quick brown fox jumps over the lazy dog. The
      quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.The quick brown fox jumps
      over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.The
      quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps
      over the lazy dog.
    </Text>
  )
}
export const NoTooltipWhenThereIsEnoughSpaceToRenderTheWholeText: Story<TextProps> = args => {
  return (
    <Text font="normal" width={300} lineClamp={3} {...args}>
      The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox
      jumps over the lazy dog.
    </Text>
  )
}
export const LeftIcon: Story<TextProps> = args => {
  return (
    <>
      <Text inline icon="service-datadog" iconProps={{ size: 20 }} {...args}>
        Datadog
      </Text>
      <Text inline icon="full-circle" iconProps={{ size: 10, color: Color.BLUE_500 }} {...args}>
        qa-cluster
      </Text>
      <Text inline icon="cross" iconProps={{ size: 16, color: Color.RED_500 }} {...args}>
        Disconnected
      </Text>
    </>
  )
}
export const RightIcon: Story<TextProps> = args => {
  return (
    <>
      <Text inline rightIcon="stop" rightIconProps={{ size: 16, color: Color.BLUE_500 }} {...args}>
        2 Instances
      </Text>
      <Text inline rightIcon="tick" rightIconProps={{ size: 16, color: Color.GREEN_500 }} {...args}>
        Password must have at least 1 uppercase.
      </Text>
    </>
  )
}
export const RightAlignment: Story<TextProps> = args => {
  return (
    <>
      <Text icon="service-datadog" iconProps={{ size: 20 }} style={{ justifyContent: 'flex-end' }} {...args}>
        Datadog
      </Text>
    </>
  )
}
export const CombiningWithOtherStyledProps: Story<TextProps> = args => {
  return (
    <Layout.Vertical id="text-styled-props">
      <Text
        padding="small"
        font={{ size: 'large', align: 'center' }}
        border={{ color: 'green500' }}
        background="grey100"
        {...args}>
        The quick brown fox jumps over the lazy dog
      </Text>
    </Layout.Vertical>
  )
}
export const SupTextBasic: Story<TextProps> = args => {
  return (
    <Text {...args}>
      Continuous Efficiency
      <SupText background="yellow600" color="white">
        beta
      </SupText>
    </Text>
  )
}
