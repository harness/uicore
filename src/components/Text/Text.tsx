import React, { HTMLAttributes, useRef, useLayoutEffect, useState, useEffect } from 'react'
import { Assign } from 'utility-types'
import cx from 'classnames'
import { StyledProps, styledClasses, omitStyledProps } from '../../styled-props/StyledProps'
import styledCSS from '../../styled-props/StyledProps.css'
import { OptionalTooltip } from '../../core/Types'
import { Utils } from '../../core/Utils'
import css from './Text.css'
import { Icon, IconName, IconProps } from '../../icons/Icon'
import { HarnessDocTooltip } from '../../frameworks/Tooltip/Tooltip'

export interface TextProps extends Assign<HTMLAttributes<HTMLDivElement>, StyledProps>, OptionalTooltip {
  // When lineClamp is specified, show ... (ellipsis) when text is overflown and show the full text
  // in a tooltip on hover. Note that `tooltip` prop takes precedence over this prop.
  // See more: https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp
  lineClamp?: number

  /** When true, tooltip will always be shown (even in case lineClamp event is not triggered: no `...` is rendered)  */
  alwaysShowTooltip?: boolean

  /** Left icon */
  icon?: IconName

  /** Optional props for left icon */
  iconProps?: Partial<IconProps>

  /** Right icon */
  rightIcon?: IconName

  /** Optional props for right icon */
  rightIconProps?: Partial<IconProps>

  /** Optional HTML tag. Default is 'p' */
  tag?: string
}

export function Text(props: TextProps) {
  const Tag = (props.tag ? props.tag : props.inline ? 'span' : 'p') as React.ElementType
  const [tooltip, setTooltip] = useState(props.tooltip)
  const lineClamp = props.lineClamp
  const style = { ...props.style }
  const ref = useRef<HTMLElement>()
  const { icon, iconProps, rightIcon, rightIconProps, alwaysShowTooltip } = props

  if (lineClamp) {
    ;(style as any)['--text-line-clamp'] = lineClamp // eslint-disable-line
  }

  if (icon || rightIcon) {
    ;(style as any)['display'] = props.inline ? 'inline-flex' : 'flex' // eslint-disable-line
    ;(style as any)['alignItems'] = 'center' // eslint-disable-line
  }

  useLayoutEffect(() => {
    const renderTooltip = () => {
      const heightDifference = (ref.current?.scrollHeight || 0) - (ref.current?.offsetHeight || 0)
      const widthDifference = (ref.current?.scrollWidth || 0) - (ref.current?.offsetWidth || 0)
      // Reason to consider difference more than 1 px is
      // If height is 18.2 Px, chrome taking scrollHeight as 18px and firefox take this as 19px
      // where as offsetHeight both take it as 18px
      if (lineClamp && lineClamp > 0 && (heightDifference > 1 || widthDifference > 1)) {
        setTooltip(props.tooltip || (props.children as JSX.Element))
      } else {
        setTooltip(undefined)
      }
    }

    renderTooltip()
    addEventListener('resize', renderTooltip)

    return () => {
      removeEventListener('resize', renderTooltip)
    }
  }, [props.children, props.tooltip, props.tooltipProps])

  useEffect(() => {
    if ((!lineClamp || alwaysShowTooltip) && props.tooltip) {
      setTooltip(props.tooltip)
    }
  }, [lineClamp, props.tooltip, alwaysShowTooltip])

  const wrappedInTooltip = (
    <Utils.WrapOptionalTooltip
      tooltip={tooltip}
      tooltipProps={{
        ...props.tooltipProps,
        targetClassName: cx(css.targetClass, props.tooltipProps?.targetClassName)
      }}>
      <Tag
        {...omitStyledProps(
          Object.assign({}, props, { style }),
          'tooltip',
          'tooltipProps',
          'lineClamp',
          'icon',
          'iconProps',
          'rightIcon',
          'rightIconProps',
          'tag'
        )}
        className={styledClasses(props, styledCSS.font, lineClamp && lineClamp >= 1 && css.lineclamp)}
        ref={ref}>
        {icon && <Icon className={css.icon} name={icon} size={16} padding={{ right: 'xsmall' }} {...iconProps} />}
        {props.children}
        {rightIcon && (
          <Icon className={css.icon} name={rightIcon} size={16} padding={{ left: 'xsmall' }} {...rightIconProps} />
        )}
      </Tag>
    </Utils.WrapOptionalTooltip>
  )

  return props.tooltipProps?.dataTooltipId ? (
    <div className={css.withDocsTooltip} data-tooltip-id={props.tooltipProps?.dataTooltipId}>
      {wrappedInTooltip}
      <HarnessDocTooltip tooltipId={props.tooltipProps?.dataTooltipId} useStandAlone={true} />
    </div>
  ) : (
    wrappedInTooltip
  )
}

export const SupText: React.FC<TextProps> = props => (
  <sup>
    <Text inline font="xsmall" style={{ padding: '0 2px', marginLeft: 2 }} color="red500" {...props} />
  </sup>
)
