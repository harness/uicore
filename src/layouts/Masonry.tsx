import { Container, Spacing } from '../'
import cx from 'classnames'
import MasonryLayout from 'masonry-layout'
import React, { useEffect, useState } from 'react'
import { PaddingProps } from 'styled-props/padding/PaddingProps'

const MASONRY_UNIQUE_CLASS_PREFIX = 'uicore-masonry-layout-'

export type MasonryRef = InstanceType<typeof MasonryLayout>

export interface MasonryProps<T = Record<string, unknown>> {
  items: T[]
  renderItem: (item: T) => React.ReactElement
  keyOf: (item: T) => string | undefined
  gutter?: number
  masonryRef?: (masonry: MasonryRef) => void
  center?: boolean
  width?: string | number
  height?: string | number
  padding?: Spacing | PaddingProps
  className?: string
  style?: React.CSSProperties
}
export function Masonry<T = Record<string, unknown>>(props: MasonryProps<T>): React.ReactElement {
  const {
    items,
    renderItem,
    keyOf,
    gutter = 20,
    masonryRef,
    width = '100%',
    height = '100%',
    padding = 'xxxlarge',
    className,
    style,
    center
  } = props
  const classUniqueIndex = document.querySelectorAll(`[class*=${MASONRY_UNIQUE_CLASS_PREFIX}-container]`)?.length ?? 0
  const [containerClass] = useState(`${MASONRY_UNIQUE_CLASS_PREFIX}-container${classUniqueIndex}`)
  const [itemClass] = useState(`${MASONRY_UNIQUE_CLASS_PREFIX}-item-${classUniqueIndex}`)
  const [masonry, setMasonry] = useState<MasonryRef>()
  const _style = Object.assign({}, style, center && { margin: '0 auto' })

  useEffect(() => {
    const container = document.querySelector('.' + containerClass)

    if (container) {
      if (masonry) {
        masonry.destroy?.()
      }

      const _masonry = new MasonryLayout(container, {
        itemSelector: '.' + itemClass,
        gutter,
        fitWidth: !!center,
        transitionDuration: 0
      })

      setMasonry(_masonry)
      masonryRef?.(_masonry)
    }
  }, [items])

  return (
    <Container width={width} height={height} padding={padding} className={cx(containerClass, className)} style={_style}>
      {items?.map(item => (
        <Container className={itemClass} key={keyOf(item)} style={{ marginBottom: gutter }}>
          {renderItem(item)}
        </Container>
      ))}
    </Container>
  )
}
