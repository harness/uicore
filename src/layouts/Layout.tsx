/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { HTMLAttributes } from 'react'
import { StyledProps, styledClasses, omitStyledProps } from '../styled-props/StyledProps'
import css from './Layout.css'
import { Spacing } from 'core/Spacing'
import { Masonry, MasonryRef, MasonryProps } from './Masonry'

export interface LayoutProps extends HTMLAttributes<HTMLDivElement>, StyledProps {
  /** Spacing among children */
  spacing?: Spacing
}

function Vertical(props: LayoutProps, ref: React.Ref<HTMLDivElement>) {
  return (
    <div
      ref={ref}
      {...omitStyledProps(props)}
      className={styledClasses(props, css.vertical, css[`layout-spacing-${props.spacing}`])}>
      {props.children}
    </div>
  )
}

function Horizontal(props: LayoutProps, ref: React.Ref<HTMLDivElement>) {
  return (
    <div
      ref={ref}
      {...omitStyledProps(props)}
      className={styledClasses(props, css.horizontal, css[`layout-spacing-${props.spacing}`])}>
      {props.children}
    </div>
  )
}

const Layout = { Vertical: React.forwardRef(Vertical), Horizontal: React.forwardRef(Horizontal), Masonry }

export { Layout, MasonryRef, MasonryProps }
