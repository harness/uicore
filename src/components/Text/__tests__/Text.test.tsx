import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import { Text } from '../Text'
import { Layout } from 'layouts/Layout'

describe('Text component fixes', () => {
  test('basic rendering with nested', async () => {
    const { container, getByText, queryByText } = render(
      <Text tooltip="ABC tooltip">
        <Layout.Horizontal>
          <Text>Nested</Text>
        </Layout.Horizontal>
      </Text>
    )
    fireEvent.mouseOver(getByText('Nested'))
    await waitFor(() => expect(queryByText('ABC tooltip')).toBeTruthy())
    expect(container).toMatchSnapshot('nested text')
  })
})
