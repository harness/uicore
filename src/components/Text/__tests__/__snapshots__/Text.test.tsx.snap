// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component fixes basic rendering with nested: nested text 1`] = `
<div>
  <span
    class="bp3-popover-wrapper"
  >
    <span
      class="bp3-popover-target bp3-popover-open"
    >
      <p
        class="font main"
        tabindex="0"
      >
        <div
          class="horizontal layout-spacing-undefined main"
        >
          <p
            class="font main"
          >
            Nested
          </p>
        </div>
      </p>
    </span>
  </span>
</div>
`;
